File: src/sys/component_manager/lib/builtins/src/root_job.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use anyhow::Error;
use fidl_fuchsia_kernel as fkernel;
use fuchsia_runtime::job_default;
use futures::TryStreamExt;

/// An implementation of the `fuchsia.kernel.RootJob` protocol.
pub struct RootJob;

impl RootJob {
    pub async fn serve(
        mut stream: fkernel::RootJobRequestStream,
        rights: zx::Rights,
    ) -> Result<(), Error> {
        let job = job_default();
        while let Some(fkernel::RootJobRequest::Get { responder }) = stream.try_next().await? {
            responder.send(job.duplicate(rights)?)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use fuchsia_async as fasync;
    use futures::TryFutureExt;
    use zx::AsHandleRef;

    #[fuchsia::test]
    async fn has_correct_rights() -> Result<(), Error> {
        let (proxy, stream) = fidl::endpoints::create_proxy_and_stream::<fkernel::RootJobMarker>();
        fasync::Task::local(
            RootJob::serve(stream, zx::Rights::TRANSFER)
                .unwrap_or_else(|err| panic!("Error serving root job: {}", err)),
        )
        .detach();

        let root_job = proxy.get().await?;
        let info = zx::Handle::from(root_job).basic_info()?;
        assert_eq!(info.rights, zx::Rights::TRANSFER);
        Ok(())
    }
}

File: sdk/lib/driver_test_realm/src/root_job.cc
// Copyright 2024 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <lib/driver_test_realm/src/root_job.h>
#include <lib/syslog/cpp/macros.h>

namespace driver_test_realm {

void RootJob::Get(GetCompleter::Sync& completer) {
  zx::job job;
  zx_status_t status = zx::job::default_job()->duplicate(ZX_RIGHT_SAME_RIGHTS, &job);
  if (status != ZX_OK) {
    FX_LOG_KV(ERROR, "Failed to duplicate job", FX_KV("status", status));
  }
  completer.Reply(std::move(job));
}

}  // namespace driver_test_realm
